include/kframework/lemmas/bytes-simplification.k

```diff
@@ -7,88 +7,240 @@ module BYTES-SIMPLIFICATION [symbolic]
-      requires 0 <=Int S0 andBool 0 <=Int L0 andBool 0 <=Int S1 andBool 0 <=Int L1 [simplification]
+    rule [buf-inject-ml]:
+      { #buf(W:Int, X:Int) #Equals #buf(W:Int, Y:Int) } => { X #Equals Y }
+      requires 0 <=Int W
+       andBool #rangeUInt (8 *Int W, X)
+       andBool #rangeUInt (8 *Int W, Y)
+      [simplification]
 
-    rule WS         +Bytes .Bytes => WS  [simplification]
-    rule .Bytes +Bytes WS         => WS  [simplification]
+    rule [buf-as-int-left]:  B:Bytes   ==K #buf(32, X:Int) => X ==Int #asInteger(B) requires lengthBytes(B) <=Int 32 [simplification, concrete(B)]
+    rule [buf-as-int-right]: #buf(32, X:Int) ==K B:Bytes   => X ==Int #asInteger(B) requires lengthBytes(B) <=Int 32 [simplification, concrete(B)]
 
-    // Re-associating bytes concatenation into a canonical right-associative form
-    rule (BA1 +Bytes BA2) +Bytes BA3 => BA1 +Bytes (BA2 +Bytes BA3) [simplification]
+    rule [buf-asWord-invert]:
+      #buf (W:Int , #asWord(B:Bytes)) => #buf(W -Int lengthBytes(B), 0) +Bytes B
+      requires lengthBytes(B) <=Int W
+      [simplification]
 
-    // #range
+    rule [buf-zero-concat-base]:
+      #buf(W1:Int, 0) +Bytes #buf(W2:Int, 0) => #buf(W1 +Int W2, 0)
+      [simplification]
 
-    rule #range(_, _, K) => .Bytes requires K <=Int 0 [simplification]
+    rule [buf-zero-concat-conc]:
+      #buf(W1:Int, 0) +Bytes #buf(W2:Int, 0) +Bytes B => #buf(W1 +Int W2, 0) +Bytes B
+      [simplification]
 
-    rule #range(#writeRange(M, N, BUF:Bytes ), L, K) =>
-         #let W = minInt(K, N -Int L) #in #range(M, L, W) +Bytes #range(#writeRange(M, N, BUF), N, K -Int W)
-      requires 0 <Int K andBool L <Int N [simplification]
+    // #range
 
-    rule #range(#writeRange(M, N, BUF:Bytes), L, K) =>
-           #let S = L -Int N #in
-           #let W = minInt(K, lengthBytes(BUF) -Int S) #in
-           #range(BUF, S, W) +Bytes #range(M, N +Int lengthBytes(BUF), K -Int W)
-      requires 0 <Int K andBool N <=Int L andBool L <=Int N +Int lengthBytes(BUF) [simplification]
+    rule [range-empty]:
+      #range(_:Bytes, S:Int, W:Int) => .Bytes
+      requires S <Int 0 orBool W <=Int 0
+      [simplification(1)]
+
+    rule [range-full]:
+      #range(B:Bytes, 0, W:Int) => B
+      requires lengthBytes(B) ==Int W
+      [simplification]
+
+    rule [range-overfull]:
+      #range(B:Bytes, 0, W:Int) => B +Bytes #buf(W -Int lengthBytes(B), 0)
+      requires lengthBytes(B) <Int W
+      [simplification]
+
+    rule [range-params-equal-ml]:
+      { #range (B1:Bytes, S1:Int, W1:Int) #Equals #range (B2:Bytes, S2:Int, W2:Int) } => #Top
+      requires B1 ==K B2 andBool S1 ==Int S2 andBool W1 ==Int W2
+      [simplification]
+
+    rule [range-included-in-cHead]:
+      #range(B1:Bytes +Bytes _:Bytes, S:Int, W:Int) => #range(B1, S, W)
+      requires S +Int W <=Int lengthBytes(B1)
+      [simplification]
```

Comment:
```suggestion
    rule [range-included-in-cHead]:
      #range(B1:Bytes +Bytes _:Bytes, S:Int, W:Int) => #range(B1, S, W)
      requires 0 <=Int S andBool 0 <=Int W andBool S +Int W <=Int lengthBytes(B1)
      [simplification]
```

Simplification priorities do not work like normal priorities, they are _suggestions_, not _requirements_ for the backend. So we must write `simplification` rules that are sound regardless of the higher-priority simplification rules that are present: https://kframework.org/docs/user_manual/#simplification-attribute

Reply:
It looks like many of the `#range` rules may have this problem, make sure to check them all.

In cases where you have something ilke `lengthBytes(B) <=Int W` or such, then it's not needed because `lengthBytes` should always be positive.

Reply:
So, from what I see:

- `[range-empty]` is ok since it specifically targets the less-than-zero cases.
- `[range-full]`, `[range-overfull]`, and `[range-includes-cHead]` are ok because `S` is explicitly zero and `W` is bounded by `lengthBytes(B)`.
- `[range-params-equal-ml]` is ok because all parameters are equal on the LHS and RHS.
- `[range-included-in-cHead]` (the one you mentioned) is ok because if `S <Int 0` or `W <Int 0` on the LHS then `S <Int 0` or `W <Int 0` also on the RHS.
- `[range-outside-cHead]` is ok because `S` is bounded by `lengthBytes(B1)` and if `W <Int 0` on the LHS then also `W <Int 0` on the RHS.
- `[range-inside-cHead]` is ok because `0 <Int S` and if `W <Int 0` on the LHS then also `W <Int 0` on the RHS.
- `[range-join-base]`, `[range-join-conc]`, `[range-reform-base]`, `[range-reform-conc]`, and `[range-buf-zero-1]` are protected directly.
- `[range-writeRange-before]` and `[range-writeRange-in-between]`: added `0 <Int W2` as per the previous versions.
- `[range-writeRange-outside]` is ok because `S1` and `W2` are protected directly and `S2` is bounded by the (non-negative) expression in the `requires` clause.

`[range-buf-zero-2]`, however, is a bit of mess. It is now:
```
  rule [range-buf-zero-2]:
    #range(#buf(_:Int, 0), S:Int, W:Int) => #buf(W, 0)
    requires 0 <=Int S andBool 0 <=Int W
    [simplification]
```
but I'm not sure about the first parameter of `#buf`, given the (non-)totality of `#buf`.

Reply:
I've added a constraint on the first parameter, just to be safe.